Extend/Payment_Gateway.php

```diff
@@ -0,0 +1,266 @@
+            $oOrder->oxorder__oxbillemail->value,
+            $user->oxuser__oxfname->value,
+            $user->oxuser__oxlname->value
+        ));
+
+        $payment = oxNew(\OxidEsales\Eshop\Application\Model\Payment::class);
+        $payment->load(self::NAME);
+
+        if ($payment->oxpayments__wdoxidee_additional_info->value) {
+            $oViewConf = oxNew('oxViewConfig');
+            $ip = $oViewConf->getRemoteAddress();
+            $oTransaction->setIpAddress($ip);
+            $oTransaction->setConsumerId($user->oxuser__oxid->value);
+            $oTransaction->setOrderNumber($oOrder->oxorder__oxid->value);
+        }
+
+        if ($payment->oxpayments__wdoxidee_descriptor->value) {
+            $descriptor = self::getDescriptor($oOrder->oxorder__oxid->value);
+            $oTransaction->setDescriptor($descriptor);
+        }
+
+        $address = new \Wirecard\PaymentSdk\Entity\Address(
+            self::getCountryCode($user->oxuser__oxcountryid),
+            $user->oxuser__oxcity->value,
+            $user->oxuser__oxstreet->value
+        );
+        $address->setPostalCode($user->oxuser__oxzip->value);
+
+        $accountHolder = new \Wirecard\PaymentSdk\Entity\AccountHolder();
+        $accountHolder->setAddress($address);
+        $accountHolder->setFirstName($user->oxuser__oxfname->value);
+        $accountHolder->setLastName($user->oxuser__oxlname->value);
+        $accountHolder->setPhone($user->oxuser__oxfon->value);
+        $accountHolder->setEmail($oOrder->oxorder__oxbillemail->value);
+
+        $oTransaction->setAccountHolder($accountHolder);
+
+        if ($oOrder->oxorder__oxdelfname->value) { //shipping address exists
+            $addressShipping = new \Wirecard\PaymentSdk\Entity\Address(
+                self::getCountryCode($oOrder->oxorder__oxdelcountryid),
+                $oOrder->oxorder__oxdelcity->value,
+                $oOrder->oxorder__oxdelstreet->value
+            );
+            $addressShipping->setPostalCode($oOrder->oxorder__oxdelzip->value);
+            $accountHolderShipping = new \Wirecard\PaymentSdk\Entity\AccountHolder();
+            $accountHolderShipping->setAddress($addressShipping);
+            $accountHolderShipping->setFirstName($oOrder->oxorder__oxdelfname->value);
+            $accountHolderShipping->setLastName($oOrder->oxorder__oxdellname->value);
+            $accountHolderShipping->setPhone($oOrder->oxorder__oxdelfon->value);
+            $oTransaction->setShipping($accountHolderShipping);
+        } else {
+            $oTransaction->setShipping($accountHolder);
+        }
+
+        if ($payment->oxpayments__wdoxidee_basket->value) { //add basket data
+            $finalPrice = array();
+            $contents = $basket->getContents();
+            foreach ($contents as $content) {
+                $finalPrice[$content->getProductId()] = $content->getFUnitPrice();
+            }
+
+            $wdBasket = new \Wirecard\PaymentSdk\Entity\Basket();
```

Comment:
you are not setting the version of the basket

Reply:
Please take a look at the opencart plugin for example and adopt the code, check that you also set shipping into the basket as it is an item. Don't forget the taxes ;)

https://github.com/wirecard/opencart-ee/blob/64b1d8fba8f6f079e173259f9ea06c8197a8eb55/catalog/model/extension/payment/wirecard_pg/helper/pg_basket.php

Reply:
will do so in the refinement of the 'additional info'. that will be another PR.

Reply:
If you set the data, set it correctly. Or don't set it at all.

Reply:
if the version is a mandatory field it should be passed in the constructor to state that it is mandatory. how would we know that this is mandatory? also the shipping and taxes?

Reply:
so taxes should only be set when payment method is `ratepayinvoice`?

Reply:
As said in the workshop look at the implementation of paymentSDK/examples and at opencart